straxen/plugins/peak_processing.py

```diff
@@ -331,98 +336,393 @@ def find_n_competing(peaks, windows, fraction):
+            # Only time shadow gives the nearest large peak
+            if 'time' in key:
+                dtype.append(((f'time difference to the nearest previous large {type_str}', 
+                               f'nearest_dt_{type_str}'), np.int64))
+        # Also record the PDF of HalfCauchy when calculating S2 position shadow
+        s2_position_shadow_dtype.append((('PDF describing correlation to the previous large s2', 
+                                          'pdf_s2_position_shadow'), np.float32))
+
+        dtype = s1_time_shadow_dtype + s2_time_shadow_dtype + s2_position_shadow_dtype + nearest_dtype + strax.time_fields
+        return dtype
+
+    @property
+    def shadowdtype(self):
+        dtype = []
+        dtype += [('shadow', np.float32), ('dt', np.int64)]
+        dtype += [('x', np.float32), ('y', np.float32)]
+        dtype += [('nearest_dt', np.int64)]
         return dtype
 
     def compute(self, peaks):
-        roi_shadow = np.zeros(len(peaks), dtype=strax.time_fields)
-        roi_shadow['time'] = peaks['center_time'] - self.time_window_backward
-        roi_shadow['endtime'] = peaks['center_time'] - self.n_drift_time
-
-        mask_pre_s2 = peaks['area'] > self.s2_threshold
-        mask_pre_s2 &= peaks['type'] == 2
-        split_peaks = strax.touching_windows(peaks[mask_pre_s2], roi_shadow)
-        res = np.zeros(len(peaks), self.dtype)
-        res['pre_s2_x'] = np.nan
-        res['pre_s2_y'] = np.nan
-        if len(peaks):
-            self.compute_shadow(peaks, peaks[mask_pre_s2], split_peaks, self.exponent, res)
-
-        res['time'] = peaks['time']
-        res['endtime'] = strax.endtime(peaks)
-        return res
+        return self.compute_shadow(peaks, peaks)
+
+    def compute_shadow(self, peaks, current_peak):
+        # 1. Define time window for each peak, we will find previous peaks within these time windows
+        roi_shadow = np.zeros(len(current_peak), dtype=strax.time_fields)
+        roi_shadow['time'] = current_peak['center_time'] - self.config['shadow_time_window_backward']
+        roi_shadow['endtime'] = current_peak['center_time']
+
+        # 2. Calculate S2 position shadow, S2 time shadow, and S1 time shadow
+        result = np.zeros(len(current_peak), self.dtype)
+        for key, is_position in zip(['s2_position_shadow', 's2_time_shadow', 's1_time_shadow'], [True, False, False]):
+            type_str = key.split('_')[0]
+            stype = 2 if 's2' in key else 1
+            mask_pre = (peaks['type'] == stype) & (peaks['area'] > self.config['shadow_threshold'][key])
+            split_peaks = strax.touching_windows(peaks[mask_pre], roi_shadow)
+            array = np.zeros(len(current_peak), np.dtype(self.shadowdtype))
+
+            # Initialization
+            array['x'] = np.nan
+            array['y'] = np.nan
+            array['dt'] = self.config['shadow_time_window_backward']
+            # Shadow is set to be the lowest possible value
```

Comment:
`# Shadow is set to be the lowest possible value -> # The default value for shadow...`
 is not it?